docs/testplan/transceiver_onboarding_test_plan.md

```diff
@@ -221,14 +267,188 @@ The following tests aim to validate various functionalities of the transceiver u
+- Use the `GENERIC_N_END` placeholder for cable lengths as described in the normalization rules
+
+**Example Directory Structure:**
+
+```
+/tmp/cmis_cdb_firmware/
+├── ACMECORP/
+│   └── QSFP-100G-AOC-GENERIC_2_ENDM/
+│       ├── ACMECORP_QSFP-100G-AOC-GENERIC_2_ENDM_1.2.3.bin
+│       └── ACMECORP_QSFP-100G-AOC-GENERIC_2_ENDM_1.2.4.bin
+├── EXAMPLE_INC/
+│   └── QSFP_200G_LR4/
+│       └── EXAMPLE_INC_QSFP_200G_LR4_2.0.1.bin
+└── ...
+```
+
+###### 1.4.1.4 Firmware Binary Storage on Remote Server
+
+The CMIS CDB firmware binaries must be stored on a remote server with the following requirements:
+
+**Server Organization:**
+
+- Directory structure should mirror the SONiC device structure described above
+- Server must be accessible via HTTP/HTTPS protocols
+- Base URL configuration is defined in `cmis_cdb_firmware_base_url.csv`
+
+**Base URL Configuration:**
+The `cmis_cdb_firmware_base_url.csv` file contains the mapping between inventory files and their corresponding firmware download URLs:
+
+```csv
+inv_name,fw_base_url
+lab,http://firmware-server.example.com/cmis_cdb_firmware/
+production,https://secure-firmware.example.com/cmis_cdb_firmware/
+```
+
+**Download URL Format:**
+Firmware binaries are accessed using the following URL pattern:
+```
+<fw_base_url>/<NORMALIZED_VENDOR_NAME>/<NORMALIZED_VENDOR_PART_NUMBER>/<FIRMWARE_BINARY_NAME>
+```
+
+**Example:**
+```
+http://firmware-server.example.com/cmis_cdb_firmware/ACMECORP/QSFP-100G-AOC-GENERIC_2_ENDM/ACMECORP_QSFP-100G-AOC-GENERIC_2_ENDM_1.2.4.bin
+```
+
+##### 1.4.2 CMIS CDB Firmware Copy to DUT via sonic-mgmt infrastructure
+
+This section describes the automated process for copying firmware binaries to the DUT, ensuring only the required firmware versions are present for testing.
+
+**Firmware Selection Algorithm:**
+
+To ensure only the necessary firmware binaries are present for each transceiver:
+
+1. **Parse `transceiver_firmware_info.csv`** to obtain the list of available firmware binaries, their versions, and associated vendor and part numbers.
+2. **Parse `transceiver_dut_info.csv`** to identify the transceivers present on each DUT.
+3. **Parse `transceiver_common_attributes.csv`** to get the gold firmware version for each transceiver type.
+4. **For each unique combination of normalized vendor name and normalized part number on the DUT**, perform version sorting and selection:
+   - Parse firmware versions using semantic versioning (X.Y.Z format)
+   - Sort available firmware versions in descending order (most recent first)
+   - **Selection criteria:**
+     - Always include the gold firmware version (from `transceiver_common_attributes.csv`)
+     - Include the two most recent firmware versions
+     - If the gold firmware version is not among the two most recent versions, include it as a third binary
+     - This ensures at least 2 firmware versions are available for upgrade testing, with the gold version guaranteed to be present
```

Comment:
@mihirpat1 can the test proceed without having two different version binary files? Its not clear.

Reply:
@prgeor Reworded this to ensure that there should be at least 2 firmware binaries (in addition to the gold firmware).